File: internal/broker/self_fence.go
package broker

import "time"

// DefaultHeartbeatTimeout is the upper bound on staleness before this broker
// stops acking writes. The plan calls 3s "the load-bearing correctness
// guarantee" — combined with epoch-tagged segments, it makes the takeover
// safety delay short regardless of NFS or storage weirdness.
const DefaultHeartbeatTimeout = 3 * time.Second

// IsHeartbeatFresh reports whether the most recent message from the
// controller is within the timeout window. The Coordinator wires this onto
// the produce hot path so a broker that has lost connectivity to the
// controller stops acking writes within heartbeatTimeout.
func IsHeartbeatFresh(lastReceived time.Time, timeout time.Duration) bool {
	if lastReceived.IsZero() {
		// No heartbeat ever observed. On a fresh-broker startup this is the
		// transient state until the first ControllerCommand arrives — return
		// false so the broker rejects writes until the channel is alive.
		return false
	}
	return time.Since(lastReceived) <= timeout
}

// IsHeartbeatFresh on the Coordinator uses the default timeout. Configurable
// only via the package-level constant for now; if operators need to tune
// (e.g. on slow NFS / cross-AZ heartbeats) we'll expose a Config struct.
func (c *Coordinator) IsHeartbeatFresh() bool {
	return IsHeartbeatFresh(c.LastHeartbeat(), DefaultHeartbeatTimeout)
}

File: crates/kaas-broker/src/self_fence.rs
//! Heartbeat staleness check used by the produce hot path.
//!
//! Coordinator
//! wires this onto the Produce path so a broker that has lost
//! connectivity to the controller stops acking writes within
//! `DEFAULT_HEARTBEAT_TIMEOUT`. Combined with epoch-tagged segments,
//! this caps the takeover safety delay regardless of NFS / storage
//! weirdness.

use std::time::Duration;

use tokio::time::Instant;

/// Upper bound on staleness before this broker stops acking writes.
/// The Phase-0 plan calls 3 s "the load-bearing correctness
/// guarantee" — short enough that a controller failover doesn't
/// strand the produce path on a stale broker for long.
pub const DEFAULT_HEARTBEAT_TIMEOUT: Duration = Duration::from_secs(3);

/// Is the most recent message from the controller within the
/// staleness window? `last_received == None` means "no heartbeat
/// ever observed" — return `false` so a fresh-boot broker rejects
/// writes until the heartbeat stream is alive.
pub fn is_heartbeat_fresh(last_received: Option<Instant>, timeout: Duration) -> bool {
    match last_received {
        None => false,
        Some(t) => Instant::now().duration_since(t) <= timeout,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test(start_paused = true)]
    async fn no_heartbeat_is_not_fresh() {
        assert!(!is_heartbeat_fresh(None, DEFAULT_HEARTBEAT_TIMEOUT));
    }

    #[tokio::test(start_paused = true)]
    async fn recent_heartbeat_is_fresh() {
        let t = Instant::now();
        tokio::time::advance(Duration::from_millis(500)).await;
        assert!(is_heartbeat_fresh(Some(t), DEFAULT_HEARTBEAT_TIMEOUT));
    }

    #[tokio::test(start_paused = true)]
    async fn stale_heartbeat_is_not_fresh() {
        let t = Instant::now();
        tokio::time::advance(DEFAULT_HEARTBEAT_TIMEOUT + Duration::from_secs(1)).await;
        assert!(!is_heartbeat_fresh(Some(t), DEFAULT_HEARTBEAT_TIMEOUT));
    }
}
